javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -0,0 +1,178 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import java.io.Serializable;
+
+import static javaslang.collection.Arrays.*;
+import static javaslang.collection.NodeModifier.copyLeaf;
+import static javaslang.collection.NodeModifier.updateLeafWith;
+
+/**
+ * A `bit-mapped trie` is a very wide and shallow tree (for integer indices the depth will be ≤ 6).
+ * Each node has a maximum of `32` children (configurable).
+ * Access to a given position is done by converting the index to a base 32 number and using each digit to descend down the tree.
+ * Modifying the tree is done similarly, but along the way the path is copied, returning a new root every time.
+ * Append inserts in the last leaf, or if the tree is full from the right, it adds another layer on top of it (the old root will be the first of the new one).
+ * Prepend is done similarly, but an offset is needed, because adding a new top node (where the current root would be the last node of the new root)
+ * shifts the indices by half of the current tree's full size. The `offset` shifts them back to the correct index.
+ *
+ * @author Pap Lőrinc
+ * @since 2.1.0
+ */
+final class BitMappedTrie<T> implements Serializable {
```

Comment:
Sure, will do, but will have to delete it later when the `leading` and `trailing` are introduced to optimize `append` and `prepend` (in order to be able to iterate over all arrays, cannot combine 3 iterators efficiently otherwise)